misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt

```diff
@@ -0,0 +1,147 @@
+package misk.hibernate
+
+import com.google.common.base.Stopwatch
+import misk.jdbc.DataSourceConfig
+import misk.logging.getLogger
+import misk.resources.ResourceLoader
+import org.hibernate.SessionFactory
+import org.hibernate.query.Query
+import java.sql.Connection
+import java.util.Collections
+import java.util.NavigableMap
+import java.util.NavigableSet
+import java.util.TreeMap
+import java.util.TreeSet
+import java.util.regex.Pattern
+import javax.persistence.PersistenceException
+
+private val logger = getLogger<SchemaMigrator>()
+
+/**
+ * Manages **available** and **applied** schema migrations.
+ *
+ * Available schema migrations are SQL files in the datasource's `migrations_path` directory. Each
+ * file should contain SQL statements terminated by a `;`. The files should be named like
+ * `v100__exemplar.sql` with a `v`, an integer version, two underscores, a description, and the
```

Comment:
is this a tradition we want to carry forward?

Reply:
I think it’s good enough!

Reply:
Any reason why version rather than timestamp?

Reply:
Timestamp in milliseconds would preserve some form of order while still providing for super low probability of conflicts. The only drawback I can see is that the file name becomes rather long!